tck/features/ListOperations.feature

```diff
@@ -0,0 +1,748 @@
+  Scenario: Concatenating and returning the size of literal lists
+    When executing query:
+      """
+      RETURN size([[], []] + [[]]) AS l
+      """
+    Then the result should be:
+      | l |
+      | 3 |
+    And no side effects
+
+  Scenario: Returning nested expressions based on list property
+    And having executed:
+      """
+      CREATE (:Label)
+      """
+    When executing query:
+      """
+      MATCH (n:Label)
+      SET n.array = [1, 2, 3, 4, 5]
+      RETURN tail(tail(n.array))
+      """
+    Then the result should be:
+      | tail(tail(n.array)) |
+      | [3, 4, 5]           |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Indexing into nested literal lists
+    When executing query:
+      """
+      RETURN [[1]][0][0]
+      """
+    Then the result should be:
+      | [[1]][0][0] |
+      | 1           |
+    And no side effects
+
+  Scenario: Concatenating lists of same type
+    When executing query:
+      """
+      RETURN [1, 10, 100] + [4, 5] AS foo
+      """
+    Then the result should be:
+      | foo                |
+      | [1, 10, 100, 4, 5] |
+    And no side effects
+
+  Scenario: Appending lists of same type
+    When executing query:
+      """
+      RETURN [false, true] + false AS foo
+      """
+    Then the result should be:
+      | foo                  |
+      | [false, true, false] |
+    And no side effects
+
+  Scenario: Execute n[0]
+    When executing query:
+      """
+      RETURN [1, 2, 3][0] AS value
+      """
+    Then the result should be:
+      | value |
+      | 1     |
+    And no side effects
+
+  Scenario: Extract eagerly
+    And having executed:
+      """
+      CREATE (:Label1 {name: 'original'})
+      """
+    When executing query:
+      """
+      MATCH (a:Label1)
+      WITH collect(a) AS nodes
+      WITH nodes, extract(x IN nodes | x.name) AS oldNames
+      UNWIND nodes AS n
+      SET n.name = 'newName'
+      RETURN n.name, oldNames
+      """
+    Then the result should be:
+      | n.name    | oldNames   |
+      | 'newName' | ['original'] |
+    And the side effects should be:
+      | +properties | 1 |
+      | -properties | 1 |
+
+  Scenario: Filter eagerly
+    And having executed:
+      """
+      CREATE (:Label1 {name: 'original'})
+      """
+    When executing query:
+      """
+      MATCH (a:Label1)
+      WITH collect(a) AS nodes
+      WITH nodes, filter(x IN nodes WHERE x.name = 'original') AS noopFiltered
+      UNWIND nodes AS n
+      SET n.name = 'newName'
+      RETURN n.name, length(noopFiltered)
+      """
+    Then the result should be:
+      | n.name    | length(noopFiltered)   |
+      | 'newName' | 1                      |
+    And the side effects should be:
+      | +properties | 1 |
+      | -properties | 1 |
+
+  Scenario: Length on filter
+    When executing query:
+      """
+      MATCH (n)
+      OPTIONAL MATCH (n)-[r]->(m)
+      RETURN length(filter(x IN collect(r) WHERE x <> null)) AS cn
```

Comment:
`length()` should not be used for lists; we should use `size()`.
`filter()` should be replaced with a list comprehension.